pypulseq/Sequence/read_seq.py

```diff
@@ -231,36 +230,37 @@ def read(self, path: str, detect_rf_use: bool = False) -> None:
                         abs(self.grad_library.data[i][0]) == 0
                         and self.grad_library.data[i][2] > 0
                     ):
-                        self.grad_library.data[i][2] -= self.grad_raster_time
-                        self.grad_library.data[i][1] = self.grad_raster_time
+                        d = self.grad_library.data[i]
+                        self.grad_library.update(i, None, (d[0], self.grad_raster_time, d[2] - self.grad_raster_time) + d[3:], self.grad_library.type[i])
 
                 if self.grad_library.data[i][3] == 0:
                     if (
                         abs(self.grad_library.data[i][0]) == 0
                         and self.grad_library.data[i][2] > 0
                     ):
-                        self.grad_library.data[i][2] -= self.grad_raster_time
-                        self.grad_library.data[i][3] = self.grad_raster_time
+                        d = self.grad_library.data[i]
+                        self.grad_library.update(i, None, d[:2] + (d[2] - self.grad_raster_time, self.grad_raster_time) + d[4:], self.grad_library.type[i])
 
             if self.grad_library.type[i] == "g":
-                self.grad_library.data[i] = [
+                self.grad_library.update(i, None, (
                     self.grad_library.data[i][:2],
                     0,
                     self.grad_library.data[i][2:],
-                ]
-                self.grad_library.lengths[i] += 1
+                ), self.grad_library.type[i])
 
         # For versions prior to 1.4.0 block_durations have not been initialized
         self.block_durations = dict()
         # Scan through blocks and calculate durations
         for block_counter in self.block_events:
-            block = self.get_block(block_counter)
+            # Insert delay as temporary block_duration
+            self.block_durations[block_counter] = 0
             if delay_ind_temp[block_counter] > 0:
-                block.delay = SimpleNamespace()
-                block.delay.type = "delay"
-                block.delay.delay = temp_delay_library.data[
-                    delay_ind_temp[block_counter]
-                ]
+                self.block_durations[block_counter] = temp_delay_library.data[
+                        delay_ind_temp[block_counter]
+                    ][0]

```

Comment:
is this needed?

Reply:
Yes, it is needed. In the old format, delay events were explicitly defined in the event table. In the new format, only the block duration is stored in the event table. However, to calculate the duration, we need the maximum of the stored delay and the actual duration of the block events (determined with `calc_duration`). `get_block` gives us the stored block duration (where you previously got an error), i.e. the delay, which is used by `calc_duration` (but the output of `calc_duration` might be larger depending on the events).

I do feel the whole handling of the delay event is a bit messy in the new format. We define a delay event with `make_delay`, and can use it in `calc_duration`. But as soon as we add it to the sequence, the delay event disappears. Instead, `get_block` will just give us a `block_duration` field, which is a special case in `calc_duration` (`isinstance(event, (float, int))`). As a consequence, the only reason `calc_duration` can determine the duration in this case, is because the duration was already stored.

Reply:
Ah, thanks for the nice explanation. Wasn't aware that the delay event is gone after adding it to a block.